### PR TITLE
Fix architecture specific msbuild path

### PR DESCRIPTION
**Pull request by heng-liu**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/2281

Regression? No

## Description
**Context**: NuGet.exe pack  failed in Developer Command prompt from VS2022, for `Could not load file or assembly 'file:///D:\ProgramFiles\Microsoft VisualStudio\2022\Preview\MSBuild\Current\Bin\arm64\Microsoft.Build.Framework.dll'
or one of its dependencies. The system cannot find the file specified.`  The reason is, NuGet tries to get the directory of msbuild.exe from [PATH env var ](https://github.com/NuGet/NuGet.Client/blob/f1e67ed45119f3906621516d134e026e7d25bc9c/src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs#L496), and in Developer Command prompt, the path of a specific architecture name instead of a general one, is added into the PATH. 
However, msbuild doesn't plan to ship all of the MSBuild assemblies and their dependencies into the architecture specific folder(bin\arm64, bin \amd64). The copies one folder up in bin are the ones that get loaded in Visual Studio and in MSBuild.exe.

This PR is to get the general folder which contains msbuild.exe, if an architecture specific folder is found in the PATH env var.

Manually tested on an arm64 machine:
NuGet.exe 6.6.1
```
PS D:\Users\source\repos\ClassLibrary1> D:\Users\Downloads\nuget.exe pack .\ClassLibrary1\ClassLibrary1.csproj
Attempting to build package from 'ClassLibrary1.csproj'.
MSBuild auto-detection: using msbuild version '17.7.0.27901' from 'D:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\arm64'.
Could not load file or assembly 'file:///D:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\arm64\Microsoft.Build.Framework.dll' or one of its dependencies. The system cannot find the file specified.
```
The NuGet.exe from artifacts with this fix:
```
PS D:\Users\source\repos\ClassLibrary1> D:\Users\Downloads\fix\nuget.exe pack .\ClassLibrary1\ClassLibrary1.csproj
Attempting to build package from 'ClassLibrary1.csproj'.
MSBuild auto-detection: using msbuild version '17.7.0.27901' from 'D:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin'.
Packing files from 'D:\Users\source\repos\ClassLibrary1\ClassLibrary1\bin\Debug'.
WARNING: NU5115: Description was not specified. Using 'Description'.
The package ClassLibrary1.1.0.0 is missing a readme. Go to https://aka.ms/nuget/authoring-best-practices/readme to learn why package readmes are important.
Successfully created package 'D:\Users\source\repos\ClassLibrary1\ClassLibrary1.1.0.0.nupkg'.
```

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs]: Wouldn't it be better to skip the test? I think this will show up as a successful test run, which is kind of misleading if it's supposed to be not-applicable.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]:
```suggestion
    <value>The resolved MSBuild directory is `{0}` which is an architecture-specific directory. Could not find MSBuild in its parent directory (non-architecture specific). </value>
```

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: It would be better to cache the `DirectoryInfo` object so that it can used in the next line.

**Review by nkolev92 — DISMISSED** [review]
LGTM. 
I know @kartheekp-ms has a good suggestion that should be addressed, so the new push will dismiss this approval.

**Comment by donnie-msft** [comment]
> LGTM. I know @kartheekp-ms has a good suggestion that should be addressed, so the new push will dismiss this approval.

@nkolev92 Just curious about this callout & approval. So, you're OK without my suggestion for fixing typos in the error message?

**Comment by nkolev92** [comment]
There's no history to show when it was added, but if you look at your comment, I upvoted it as well. 

It was merely a single callout that there's stuff to be worked on. GH doesn't have it, but it's an approve with suggestions. I didn't think to call out every single comment there.

**Comment by donnie-msft** [comment]
> There's no history to show when it was added, but if you look at your comment, I upvoted it as well.

I don't understand the concept of "Approved" yet don't actually accept the approval. I saw the upvote, but that doesn't have anything to do with approvals.

**Review by donnie-msft — CHANGES_REQUESTED** [review]
Please fix the typos I suggested. Thanks

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx
Comment (heng-liu) [reply src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx]: Thanks for reviewing! Fixed the message~

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs
Comment (heng-liu) [reply src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: Thanks for reviewing! Fixed.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs
Comment (heng-liu) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs]: Thanks for @rainersigwald's input : the PATH can be set by any of the subscripts that the developer command prompt script calls, owned by different component teams of VS. For this particular case we have [a script that sets the MSBuild path up](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fdevdiv%2FDevDiv%2F_git%2FVS%3Fpath%3D%2Fsrc%2Fvc%2Fprojbld%2Fvcvars%2Fvsdevcmd%2Fcore%2Fmsbuild.bat.tt%26version%3DGBmain%26line%3D12%26lineEnd%3D12%26lineStartColumn%3D54%26lineEndColumn%3D59%26lineStyle%3Dplain%26_a%3Dcontents&data=05%7C01%7CLiu.Heng%40microsoft.com%7Cd4646232bd0c4e2feeaa08db742e7464%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638231515113561572%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=j8S4Y2tCEKvZb%2BIWHLjFt0XjLb0%2B73fNRR%2Bjt4Apyag%3D&reserved=0).
As is today, you can safely cut “arm64” or “amd64” from the end. There may be a time in the future when the MSBuild on PATH is a .NET Core MSBuild and then the relative path would be different.

So my understanding is, we're not 100% sure what would be the changes in the future, so I hope this could catch the runtime issue just in case, when the PATH changed and no msbuild.exe in the non-architecture specific folder. I'm not sure if we will need this or not in the future.

**Review by donnie-msft — DISMISSED** [review]

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs
Comment (donnie-msft) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs]: I was thinking along the lines of our other tests which, rather than `return`, actually pass a SkipMono attribute to the test.
https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/PlatformXunitAttributes/PlatformFactAttribute.cs#L60

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs:557]: The method name is bit confusing. I think `GetPlatformNeutralMsBuildPath` suits better because the method retrieves the path for the platform-neutral (non-architecture-specific) version of msbuild.exe.

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]:
```suggestion
            var directory = new DirectoryInfo(directory);
            var directoryName = directory.Name;
            var parentDirectory = directory.Parent.FullName;
```

**Review by kartheekp-ms — DISMISSED** [review]
Can we create DirectoryInfo object only once?

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs
Comment (heng-liu) [reply src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: Thanks! Fixed.

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs
Comment (heng-liu) [reply src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs:557]: Thanks for reviewing! I think "architecture" is a more specific word, as msbuild uses `arm64` and `amd64` as the folder name, and they are processor architectures. 
While platform has a more general word, it can also stand for OS, and some other meanings.
On the other hand, since this method is not an API, we can safely change the name if more people like other naming, in the future, considering it's kind of urgent to merge this PR. What do you think?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs
Comment (heng-liu) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs]: Sorry that I misunderstood the feedback.. Fixed by skipping mono test in a better way. Thanks!

**Review by kartheekp-ms — APPROVED** [review]

**Review by donnie-msft — APPROVED** [review]

**Pull request merged**
